src/d3a/models/market/market_structures.py

```diff
@@ -211,16 +203,16 @@ def __eq__(self, other):
                 self.requirements == other.requirements)
 
 
-def offer_or_bid_from_json_string(offer_or_bid, current_time=None):
+def offer_or_bid_from_json_string(offer_or_bid, current_time=None) -> Union[Offer, Bid]:
```

Comment:
I think that this would be a good candidate to be a class method for `BaseBidOffer`:
```python
BaseBidOffer.from_json()
```
(Also, no need to specify the `_string` part in `from_json_string`, I think).